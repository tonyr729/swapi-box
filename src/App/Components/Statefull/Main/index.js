import React, { Component } from 'react';
import './index.css';

class Main extends Component {
  constructor() {
    super()
    this.state = {
      
    }
  }

  render() {
    return(
      <div>
        <h1>SWAPI-Box</h1>
        <button>View Favorites</button>
        <button>People</button>
        <button>Planets</button>
        <button>Vehicles</button>
        {/* <CardContainer /> */}
      </div>
    )
  }
}

export default Main;