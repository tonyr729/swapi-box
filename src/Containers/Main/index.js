import React, { Component } from 'react';
import Header from '../../Components/Stateless/Header'
import CardContainer from '../CardContainer/CardContainer'
import APIManager from '../../Helpers/API/API Manager/apiManager'
import './index.css';

class Main extends Component {
  constructor() {
    super()
    this.state = {
      people: [],
      vehicles: [],
      planets: [],
      favorites: [], 
      displayed: []
    }
    this.api = new APIManager();
  }
  
  setPeopleData= async (category) => {
    if(!this.state.people.length) {
      const people = await this.api.fetchPeople();
      this.setState({
        people: people
      })
    } 
    this.setDisplayedData('people')
  }

  setVehicleData= async () => {
    if(!this.state.vehicles.length) {
      const vehicles = await this.api.fetchVehicles();
      this.setState({
        vehicles: vehicles
      })
    }
    this.setDisplayedData('vehicles')
  }

  setPlanetData = async () => {
    if(!this.state.planets.length) {
      const planets = await this.api.fetchPlanets();
      this.setState({
        planets: planets
      })
    }
    this.setDisplayedData('planets');
  }

  setDisplayedData = (category) => {
    this.setState({
      displayed: this.state[category]
    })
  }

  render() {
    return(
      <div className="main">
        <Header setPeopleData={ this.setPeopleData } setVehicleData={ this.setVehicleData } setPlanetData={ this.setPlanetData } favorites={ this.state.favorites.length } />
        <CardContainer data={ this.state.displayed } />
      </div>
    )
  }
}

export default Main;