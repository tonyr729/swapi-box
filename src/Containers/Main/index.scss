.main {
  position: absolute;
  z-index: 100;
  // animation: main 40s;
  opacity: 1;
}

@keyframes main {
  0% {
    opacity: 0;
  }
  98% {
    opacity: 0;
  }
  100% {
    opacity: 1;
  }
}