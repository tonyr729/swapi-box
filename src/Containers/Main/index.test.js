import React from 'react';
import ReactDOM from 'react-dom';
import APIManager from '../../Helpers/API/API Manager/apiManager';
import Main from './index'
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';
jest.mock('../../Helpers/API/API Manager/apiManager', () => {
  return jest.fn().mockImplementation(() => {
    return {
      fetchPeople: jest.fn().mockImplementation(() => Promise.resolve(['card1', 'card2'])),
      fetchVehicles: jest.fn().mockImplementation(() => Promise.resolve(['card1', 'card2'])),
      fetchPlanets: jest.fn().mockImplementation(() => Promise.resolve(['card1', 'card2']))
    };
  });
});

describe('Main', () => {

  describe('Component', () => {
    it('should match snapshot', () => {
      const main = renderer.create(<Main />).toJSON();

      expect(main).toMatchSnapshot();
    })
  })

  describe('setPeopleData method', () => {
    it('should call fetchPeople method of the class instance', async () => {
      const main = shallow(<Main />);
      
      main.instance().setPeopleData()

      expect(main.instance().api.fetchPeople).toHaveBeenCalled();     
      
    })

    it('should set state to the evaluation of the fetchPeople method', async () => {
      const main = shallow(<Main />);
      const expected = ['card1', 'card2']
      
      expect(main.state('displayed')).toEqual([])

      Promise.resolve(main.instance().setPeopleData())
        .then(() => main.update())
        .then(() => expect(main.state('displayed')).toEqual(expected))
    })

    it('should not call fetchPeople if state is already populated with people', async () => {
      const main = shallow(<Main />);

      main.setState({
        people: [{name: "Tony"}]
      })

      Promise.resolve(main.instance().setPeopleData())
        .then(() => main.update())
        .then(() => expect(main.instance().api.fetchPeople).not.toHaveBeenCalled())
    })
  })

  describe('setVehicleData method', () => {
    it('should call fetchVehicle method of the class instance', async () => {
      const main = shallow(<Main />);
      
      main.instance().setVehicleData()

      expect(main.instance().api.fetchVehicles).toHaveBeenCalled();     
      
    })

    it('should set state to the evaluation of the fetchVehicles method', async () => {
      const main = shallow(<Main />);
      const expected = ['card1', 'card2']
      
      expect(main.state('displayed')).toEqual([])

      Promise.resolve(main.instance().setVehicleData())
        .then(() => main.update())
        .then(() => expect(main.state('displayed')).toEqual(expected))
    })

    it('should not call fetchVehicle if state is already populated with vehicles', async () => {
      const main = shallow(<Main />);

      main.setState({
        vehicles: [{name: "Tony"}]
      })

      Promise.resolve(main.instance().setVehicleData())
        .then(() => main.update())
        .then(() => expect(main.instance().api.fetchVehicles).not.toHaveBeenCalled())
    })
  })

  describe('setPlanetData method', () => {
    it('should call fetchPlanet method of the class instance', async () => {
      const main = shallow(<Main />);
      
      main.instance().setPlanetData()

      expect(main.instance().api.fetchPlanets).toHaveBeenCalled();     
      
    })

    it('should set state to the evaluation of the fetchPlanets method', async () => {
      const main = shallow(<Main />);
      const expected = ['card1', 'card2']
      
      expect(main.state('displayed')).toEqual([])

      Promise.resolve(main.instance().setPlanetData())
        .then(() => main.update())
        .then(() => expect(main.state('displayed')).toEqual(expected))
    })

    it('should not call fetchPlanets if state is already populated with Planet', async () => {
      const main = shallow(<Main />);

      main.setState({
        planets: [{name: "Tony"}]
      })

      Promise.resolve(main.instance().setPlanetData())
        .then(() => main.update())
        .then(() => expect(main.instance().api.fetchPlanets).not.toHaveBeenCalled())
    })
  })

  describe('setFavorites method', () => {
    
    it('should set state of favorites to include incoming object', async () => {
      const main = shallow(<Main />);
      const expected = {name: "tony"}
      
      expect(main.state('favorites')).toEqual([])

      Promise.resolve(main.instance().setFavorites(expected))
        .then(() => main.update())
        .then(() => expect(main.state('favorites')).toEqual([expected]))
    })

    it('should remove the object passed from state if already a favorite', async () => {
      const main = shallow(<Main />);
      const expected = {name: "tony"}

      main.setState({
        favorites: [expected]
      })
      
      expect(main.state('favorites')).toEqual([expected])

      Promise.resolve(main.instance().setFavorites(expected))
        .then(() => main.update())
        .then(() => expect(main.state('favorites')).toEqual([]))
    })
  })
})