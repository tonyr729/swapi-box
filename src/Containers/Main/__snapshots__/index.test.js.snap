// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Main Component should match snapshot 1`] = `
<div
  className="main"
>
  <div>
    <h1
      className="header-title"
    >
      SWAPI-Box
    </h1>
    <div>
      <p>
        Favorites
      </p>
      <div>
        <p>
          0
        </p>
      </div>
    </div>
    <button
      onClick={[Function]}
    >
      People
    </button>
    <button
      onClick={[Function]}
    >
      Planets
    </button>
    <button
      onClick={[Function]}
    >
      Vehicles
    </button>
  </div>
  <div
    className="card-container"
  />
</div>
`;
